specs/signature-specification.md

```diff
@@ -258,17 +258,22 @@ The signing certificate's public key algorithm and size MUST be used to determin
 
 ### Certificate Requirements
 
-The codesigning and timestamping certificates MUST meet the following requirements. These requirements are validated both at signature generation time and signature verification time, and are applied to the certificate chain in the signature envelope. These validations are independent of certificate chain validation against a trust store.
+The **codesigning** certificates MUST meet the following requirements. These requirements are validated both at signature generation time and signature verification time, and are applied to the certificate chain in the signature envelope. These validations are independent of certificate chain validation against a trust store.
+
+The **timestamping** certificates MUST meet the following requirements as well. Requirements for leaf certificate are validated both at timestamp token generation and verification time. Requirements for root and intermediate CA certificates are only validated at timestamp token verification time. This is because at generation time, the client is NOT required to have access to the full certificate chain of the Timestamp Authority (TSA). These validations are independent of certificate chain validation against a trust store.
```

Comment:
> * At present, the notation does not facilitate the reading of the certificate chain from outside the signature. Should we conduct the validation of the root and intermediate CA (Certificate Authority) certificates during both signing and verification?

@priteshbandi According to RFC 3161 and [RFC 5652](https://www.rfc-editor.org/rfc/rfc5652#section-5.1), the TSA response may not include the intermediate or root cert in their response. We could do this check for code signing cert chain because during the signing process, a `fixed` cert chain is constructed and stored under the unsigned attribute `Certificate Chain` in the envelope. Therefore, a signer could verify the cert chain without configuring any trust store or trust policy.
However, for timestamping certificates, the cert chain is formed from the cert pool (certificates field of CMS) during timestamp verification, i.e., a trust anchor is provided by the verifier and a cert chain MUST be traced to this anchor. Therefore, a tsa cert chain is not `fixed`. The signer and the verifier (can be different entities) may form different cert chains respectively (with different trust anchors or require different intermediate certs in between). In this case, even if the signer could verify the cert chain successfully, the verifier can still find the timestamp verification failed due to `no valid cert chain`. 
We do perform signing certificate validation in both sign and verify though, which is RFC 3161 compliant.

Reply:
As per today's conversation, @Two-Hearts will analyze response from commonly used tsa to check if they return cert chain as part of timestamp. If majority of them provide cert chain in signature we can impose cert-chain validation during signing.

Also, we can always support additional usecase of supplying cert chain from outside signature.

Reply:
@priteshbandi I tested some existing TSAs, listing them below:
1. `http://rfc3161timestamp.globalsign.com/advanced` and `http://timestamp.acs.microsoft.com/` are two examples that's returning the full chain in the response. I'm able to successfully sign and verify using these TSAs.
2. `http://tsa.swisssign.net` does not contain a full chain, therefore, it no longer works.
3. After enabling the cert chain validation after signing, I found that `http://timestamp.digicert.com` and `http://timestamp.apple.com/ts01` are returning certs with signature algorithm SHA1. Therefore, they did not pass Notation's check and failed the signing.

Given above, as suggested by @priteshbandi, I think we could enforce the TSA to provide a full chain in their response and enable the chain validation after signing at this point. If a user asks for a particular TSA that's not working as expected, we could provide an option for them in a future version, such as let the user provide certificates outside the timestamp token.

Reply:
>http://timestamp.digicert.com and http://timestamp.apple.com/ts01 are returning certs with signature algorithm SHA1.
What does that mean? Certs are signed using SHA1 ?

Reply:
> > http://timestamp.digicert.com and http://timestamp.apple.com/ts01 are returning certs with signature algorithm SHA1.
> > What does that mean? Certs are signed using SHA1 ?

Yes, here's the detail error for `digicert TSA`:
`Error: timestamp: root certificate with subject "CN=DigiCert Assured ID Root CA,OU=www.digicert.com,O=DigiCert Inc,C=US" is invalid or not self-signed. Certificate chain must end with a valid self-signed root certificate. Error: x509: cannot verify signature: insecure algorithm SHA1-RSA`

Reply:
Updated the spec to reflect the above change:
Timestamp cert chain is also validated during the signing process.

Reply:
Created the issue https://github.com/notaryproject/specifications/issues/302 for tracking further work according to community meeting on Jun 17 @priteshbandi @Two-Hearts

Reply:
Resolving this comment based on discussion in the 6/17/24 community meeting.